nc/source/graphics/resources/UniformBuffer.h

```diff
@@ -18,20 +19,50 @@ namespace nc::graphics
     {
         public:
             UniformBuffer();
-            UniformBuffer(nc::graphics::Graphics* graphics, const EnvironmentData& data);
+            template <typename T>
+            UniformBuffer(nc::graphics::Graphics* graphics, const T& data);
             ~UniformBuffer() noexcept;
             UniformBuffer(UniformBuffer&&);
             UniformBuffer& operator=(UniformBuffer&&);
             UniformBuffer& operator=(const UniformBuffer&) = delete;
             UniformBuffer(const UniformBuffer&) = delete;
             
             vk::Buffer* GetBuffer();
-            void Bind(nc::graphics::Graphics* graphics, const EnvironmentData& data);
+            template <typename T>
+            void Bind(nc::graphics::Graphics* graphics, const T& data);
             void Clear();
 
         private:
             Base* m_base;
             uint32_t m_memoryIndex;
             vk::Buffer m_uniformBuffer;
     };
+
+    template <typename T>
+    void UniformBuffer::Bind(Graphics* graphics, const T& data)
```

Comment:
Since we only use T for memcpy, you could just take const void* and a size here instead of const T&. Then we can avoid having to include graphics here.